proposals/0000-amending-mrp.rst

```diff
@@ -0,0 +1,256 @@
+There is also the enablement of the warnings by default on `gitlab <https://gitlab.haskell.org/ghc/ghc/-/merge_requests/3174>`_
+and its `own proposal <https://github.com/ghc-proposals/ghc-proposals/pull/314>`_.
+
+
+Proposed Change Specification
+-----------------------------
+To summarise the proposal to date:
+
+* When making ``Applicative`` a superclass of Monad, there are now redundant 
+  methods in ``Monad`` in ``return`` and ``(>>)``, as these can always (modulo
+  performance) be replaced with ``pure`` and ``(*>)`` respectively
+* The first phase was completed in 2015, that of adding a warning (off by 
+  default) that warns when definitions of ``return`` and ``(>>)`` are present
+  that aren't just making them equivalent to ``pure`` and ``(*>)`` respectively.
+  This is ``-Wnoncanonical-monad-instances``
+* In ~2020 this warning was enabled by default in 9.2.1
+
+The next phases of the proposal have not been moved forward until very recently.
+
+* Phase 2 requires that the previous warning is now an error and that 
+  ``return`` and ``(>>)`` are top level bindings, but also that canonical
+  definitions are ignored
+* Phase 3 would then start warning about canonical ``return`` and ``(>>)``
+  definitions
+* Phase 4 would remove support for any sort of override, making that an error
+
+A lawful method is simply a method that follows the laws of the typeclass.
+A canonical method is one that is always definitionally the same, and has no 
+greater power. For example, ``return`` and ``pure`` must always be equivalent, 
+and there is no reason that they should not be defined in terms of each other. 
+The proposal seeks to eliminate the presence of the excess canonical methods
+that can be defined in ``Monad``.
+
+This proposal would adjust the next phases to reduce the number of changes
+neccessary in GHC as follows:
+
+* Phase 2 makes ``-Wnoncanonical-monad-instances`` a compiler error by default,
+  and adds a ``-Wredundant-canonical-monad-method`` warning that makes canonical
+  method definitions (only ``return = pure``) warn
+* Phase 3 would then move ``return`` and ``(>>)`` to the top level, and remove
+  the original and new warnings (after removing these methods from ``Monad`` the
+  compiler will emit an error like for any other extraneous typeclass instance
+  method definition)
+
+  * Phase 3 shouldn't be  (fully) implemented until research is done as to what 
+    we want the eventual definition of ``(>>)`` to be
+
+This reduces the number of steps neccessary, and means that we wouldn't need
+some special casing about ignoring canonical definitions while we have a top
+level one.
+
+Another adjustment to the plan is to adjust the warnings and errors about
+``(>>)`` to encourage a performant definition of ``(*>)`` instead of blindly
+replacing its definition. We wish to do this because the default
+implementation of ``(*>)`` (that of ``(*>) a b = (id <$ a) <*> b``) leaks space,
```

Comment:
This is an overstatement of the issue, I think. The ‘space leak’ isn't really a leak; it's a failure to take an opportunity to do tail-call optimization. For monads for which tail-call optimization is never possible (because, for example, the monad contains more than one element), there is no ‘leak’, and no issue with the performance of the default implementation as far as I know.